ethcore/src/engines/authority_round/mod.rs

```diff
@@ -125,54 +140,96 @@ impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
+		self.opt_duration_remaining().unwrap_or_else(|| {
+			let ctr = self.load();
+			error!(target: "engine", "Step counter under- or overflow: {}, aborting", ctr);
+			panic!("step counter under- or overflow: {}", ctr)
+		})
+	}
+
+	fn opt_duration_remaining(&self) -> Option<Duration> {
 		let now = unix_now();
-		let expected_seconds = self.load()
-			.checked_add(1)
-			.and_then(|ctr| ctr.checked_mul(self.duration as u64))
-			.map(Duration::from_secs);
-
-		match expected_seconds {
-			Some(step_end) if step_end > now => step_end - now,
-			Some(_) => Duration::from_secs(0),
-			None => {
-				let ctr = self.load();
-				error!(target: "engine", "Step counter is too high: {}, aborting", ctr);
-				panic!("step counter is too high: {}", ctr)
-			},
+		let mut prev_dur = u64::from(self.durations[&0]);
+		let mut prev_step = 0u64;
+		let mut prev_time = 0u64;
+		let next_step = self.load().checked_add(1)?;
+		for (time, dur) in self.durations.iter().skip(1) {
+			let step_diff = time.checked_add(prev_dur)?.checked_sub(1)?.checked_sub(prev_time)?.checked_div(prev_dur)?;
+			let step = prev_step.checked_add(step_diff)?;
+			if step >= next_step {
+				break;
+			}
+			prev_step = step;
+			prev_time = step_diff.checked_mul(prev_dur)?.checked_add(prev_time)?;
+			prev_dur = u64::from(*dur);
+		}
+		let time = prev_time.checked_add(next_step.checked_sub(prev_step)?.checked_mul(prev_dur)?)?;
+		let step_end = Duration::from_secs(time);
+		if step_end > now {
+			Some(step_end - now)
+		} else {
+			Some(Duration::from_secs(0))
 		}
-
 	}
 
+	/// Increments the step number.
+	///
+	/// Panics if the new step number is `u64::MAX`.
 	fn increment(&self) {
-		use std::usize;
 		// fetch_add won't panic on overflow but will rather wrap
 		// around, leading to zero as the step counter, which might
 		// lead to unexpected situations, so it's better to shut down.
-		if self.inner.fetch_add(1, AtomicOrdering::SeqCst) == usize::MAX {
-			error!(target: "engine", "Step counter is too high: {}, aborting", usize::MAX);
-			panic!("step counter is too high: {}", usize::MAX);
+		if self.inner.fetch_add(1, AtomicOrdering::SeqCst) == u64::MAX {
+			error!(target: "engine", "Step counter is too high: {}, aborting", u64::MAX);
+			panic!("step counter is too high: {}", u64::MAX);
 		}
 
 	}
 
 	fn calibrate(&self) {
 		if self.calibrate {
-			let new_step = unix_now().as_secs() / (self.duration as u64);
-			self.inner.store(new_step as usize, AtomicOrdering::SeqCst);
+			if self.opt_calibrate().is_none() {
+				let ctr = self.load();
+				error!(target: "engine", "Step counter under- or overflow: {}, aborting", ctr);
+				panic!("step counter under- or overflow: {}", ctr)
+			}
 		}
 	}
 
+	fn opt_calibrate(&self) -> Option<()> {
+		let now = unix_now().as_secs();
+		let mut prev_dur = u64::from(self.durations[&0]);
+		let mut prev_step = 0u64;
+		let mut prev_time = 0u64;
+		for (time, dur) in self.durations.range(..now).skip(1) {
```

Comment:
This logic seems convoluted and duplicated, please at least refactor it to a single place.